phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapperForTest.java

```diff
@@ -0,0 +1,89 @@
+package org.apache.phoenix.mapreduce.index;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.hadoop.io.Text;
+import org.apache.phoenix.util.EnvironmentEdge;
+import org.apache.phoenix.util.EnvironmentEdgeManager;
+import org.apache.phoenix.util.QueryUtil;
+
+import java.io.IOException;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class IndexScrutinyMapperForTest extends IndexScrutinyMapper {
+
+    public static class ScrutinyTestClock extends EnvironmentEdge {
+        long initialTime;
+        long delta;
+
+        public ScrutinyTestClock(long delta) {
+            initialTime = System.currentTimeMillis() + delta;
+            this.delta = delta;
+        }
+
+        @Override
+        public long currentTime() {
+            return System.currentTimeMillis() + delta;
+        }
+    }
+
+    @Override
+    protected void processBatch(Context context)
```

Comment:
I think the method you're overriding here is too broad -- you've copied much of the logic you're testing, which makes the test less useful. (For example, if someone changes IndexScrutinyMapper.processBatch and breaks it, any tests using this mapper class will never notice.) 

My suggestion is to have a very narrow hook that just lets you insert the logic that needs to be different, and use the "real" code for everything else. Think of how HBase coprocs work, where HBase is doing most of the work, but occasionally has hooks that can be overriden for extra / different behavior. 

For example, an overridable, no-op "beforeQueryTargetTable" method in IndexScrutinyMapper would let you override here with just your current lines 53 and 54 setting up the different clock values.

Reply:
Basically, that if-block in IndexScrutinyMapper that I objected to last time around -- that's pretty close to what your overridden method in the subclass would look like. :-)

Reply:
Thanks @gjacoby126 Your comments really helped me get more insights on various things on this PR. :)